### PR TITLE
feat: add support for scripting/routines

### LINKED ISSUE #578: BigQuery: implement Routines API support
This is a tracking FR for implementing the Routines API functionality within the
BigQuery API.  This API underpins the features for persistent UDFs, as well as
Stored Procedures. These entities can be manipulated both via DDL statements 
(e.g. CREATE FUNCTION queries) and directly via the routines insertion method.

More implementation details about this API are shared directly with library
maintainers.  Please contact shollyman or tswast for questions.

### PR DESCRIPTION
**Pull request by callmehiphop**
Closes #578 🦕

- [x] Ensure the tests and linter pass
- [x] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

### CODE REVIEW

**Comment by callmehiphop** [comment]
Usage can be viewed easily via [system-tests](https://github.com/googleapis/nodejs-bigquery/pull/580/files#diff-55484aa8e87595a4f251670364f66a4bR680)

File: src/dataset.ts [hunk 1]
```diff
@@ -654,6 +788,102 @@ class Dataset extends ServiceObject {
...
+   *     dataset.getRoutines(nextQuery, manualPaginationCallback);
+   *   }
+   * }
+   *
+   * dataset.getRoutines({
+   *   autoPaginate: false
+   * }, manualPaginationCallback);
+   *
+   * @example <caption>If the callback is omitted a Promise will be returned</caption>
+   *  const [routines] = await dataset.getRoutines();
```
Comment (stephenplusplus) [inline src/dataset.ts]: Extra space here in front of `const`.

File: src/dataset.ts [hunk 2]
```diff
@@ -654,6 +788,102 @@ class Dataset extends ServiceObject {
...
+   * const {BigQuery} = require('@google-cloud/bigquery');
+   * const bigquery = new BigQuery();
+   * const dataset = bigquery.dataset('institutions');
+   *
+   * dataset.getRoutines((err, routines) => {
+   *   // routines is an array of `Routine` objects.
+   * });
+   *
+   * @example <caption>To control how many API requests are made and page through the results manually, set `autoPaginate` to `false`.</caption>
+   * function manualPaginationCallback(err, tables, nextQuery, apiResponse) {
```
Comment (stephenplusplus) [inline src/dataset.ts]:
```diff
- function manualPaginationCallback(err, tables, nextQuery, apiResponse) {
+ function manualPaginationCallback(err, routines, nextQuery, apiResponse) {
```

File: src/routine.ts [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+       * @returns {Promise<SetRoutineMetadataResponse>}
+       *
+       * @example
+       * const {BigQuery} = require('@google-cloud/bigquery');
+       * const bigquery = new BigQuery();
+       * const dataset = bigquery.dataset('my-dataset');
+       * const routine = dataset.routine('my_routine');
+       *
+       * const updates = {
+       *   description: 'The perfect description!'
```
Comment (stephenplusplus) [inline src/routine.ts]: I couldn't have described it better!

File: src/routine.ts [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+  ): void | Promise<common.SetMetadataResponse> {
+    // per the python client, it would appear that in order to update a routine
+    // you need to send the routine in its entirety, not just the updated fields
+    this.getMetadata((err: Error | null, fullMetadata: RoutineMetadata) => {
+      if (err) {
+        callback!(err);
+        return;
+      }
+
+      const updatedMetadata = extend(true, {}, metadata, fullMetadata);
```
Comment (stephenplusplus) [inline src/routine.ts]: Shouldn't the user-provided metadata win?

File: system-test/bigquery.ts [hunk 5]
```diff
@@ -676,6 +677,79 @@ describe('BigQuery', () => {
...
+      return bigquery.query(`
+        CREATE FUNCTION \`${routineId}\`(
+          arr ARRAY<STRUCT<name STRING, val INT64>>
+        ) AS (
+          (SELECT SUM(IF(elem.name = "foo",elem.val,null)) FROM UNNEST(arr) AS elem)
+        )
+      `);
+    });
+
+    it('should list all the routines', async () => {
```
Comment (stephenplusplus) [inline system-test/bigquery.ts]: This shouldn't depend on the previous tests. I think the ways I have handled this in the past is:

- Create an object in the before() hook, using the first `it` to confirm the name/other metadata was properly set. Then, in a list test, just check that >= 1 exists.

- Create an object in the list test, then confirm the list returns >= 1.

File: system-test/bigquery.ts [hunk 6]
```diff
@@ -676,6 +677,79 @@ describe('BigQuery', () => {
...
+        })
+        .on('end', () => {
+          assert.strictEqual(routines.length, 2);
+          assert.ok(routines[0] instanceof Routine);
+          done();
+        });
+    });
+
+    it('should check to see if a routine exists', async () => {
+      const routine = dataset.routine('my_routine');
```
Comment (stephenplusplus) [inline system-test/bigquery.ts]: Same idea as the other one, where we should create the routine we're checking the existence on in this same test.

Or... create the routine in the `before`, and re-use the object throughout the tests.

File: src/dataset.ts (hunk 2 above)
Comment (callmehiphop) [reply src/dataset.ts]: You caught me :[

File: src/routine.ts (hunk 4 above)
Comment (callmehiphop) [reply src/routine.ts]: Oof, yes, good catch.

**Comment by steffnay** [comment]
@googlebot I consent

**Comment by steffnay** [comment]
Heyo, @callmehiphop do you mind making the comment for the CLA, please?

**Comment by callmehiphop** [comment]
I consent

**Review by JustinBeckwith — APPROVED** [review]

**Pull request merged**